### DISCUSSION TITLE
How to declare function prototypes

### DISCUSSION THREAD

**Discussion by neilbalch** [cc65/cc65#2640] [General]
I wrote the following code, originally separated into a header and source file but now in one file for debugging purposes, and expected it to work based on my previous C/C++ knowledge:

```c
// #include "game_code.h"
// #include "util.h"
#pragma code-name ("CODE2")

#ifndef UTIL_H
#define UTIL_H
#define nop()  __asm__ ("NOP") // NO Operation opcode (http://www.6502.org/users/andre/65k/af65002/af65002opcodes.html)
#define sei()  __asm__ ("SEI") // Enable interrupts (set flag)
#define cli()  __asm__ ("CLI") // Disable interrupts (reset flag)
#endif

void reset();
void do_logic();
void fill_vram();

void reset() {
  nop();
}

void do_logic() {
  nop();
}

void fill_vram() {
  nop();
}
```

But instead, I get the following:

```
cc65 -t none -O --cpu 65sc02 -I ./include src/game_code.c -o build/src/game_code.c.s
src/game_code.c(20): Error: Conflicting types for 'reset'
src/game_code.c(24): Error: Conflicting types for 'do_logic'
src/game_code.c(28): Error: Conflicting types for 'fill_vram'
make: *** [Makefile:32: game_code.o] Error 1
```

I copied this into [godbolt.org](godbolt.org) and it compiles, so I'm led to think this might be down to something peculiar about `cc65`. Can someone please help me figure out what's wrong here? Thanks!

**Comment by polluks** [comment]
use
````
void reset(void);
void do_logic(void);
void fill_vram(void);
````

**Comment by neilbalch** [comment]
Oh! Could you explain why that is required here?

**Comment by polluks** [comment]
Years ago we changed the implementation
https://stackoverflow.com/questions/693788/is-it-better-to-use-c-void-arguments-void-foovoid-or-not-void-foo

**Comment by mrdudz** [comment]
Excuse me, but the error message is still misleading. no matter what the implementation change was, a prototype that looks exactly like the function signature should work. It's also valid C no matter what.

**Comment by neilbalch** [comment]
I agree. Should I reopen the issue then?

**Comment by mrdudz** [comment]
Indeed :) While i agree (void) is strongly recommended over an empty parameter list, its still valid C and should work. Or at least give a reasonable error :)

**Comment by spiro-trikaliotis** [comment]
> Excuse me, but the error message is still misleading. no matter what the implementation change was, a prototype that looks exactly like the function signature should work. It's also valid C no matter what.

Well, in C terms, void a(); is not a prototype.

Citing N1256 (the latest c99 draft), "6.2.1 scope of identifiers", No. 2:
"A function prototype is a declaration of a function that declares the types of its parameters."

In contrast, the "void a();" does not define the type of the parameters. This it is not a prototype.

Take for example this code:
```
void a();

int main() {
        a(3, 5, 7.3, "Hello");
        return 0;
}
```
It compiles flawlessly. In fact, if you have another compilation unit that defines a(), it will even run.
I would say this is reminiscence to pre-ANSI C, where you did not have any prototypes.

Look at the declaration "void a();" as if it were "void a(...);". IIRC, this is not completely correct, but comes near to what a C compiler does.

AFAIK, compilers tend to do one of the following the help the user if you use a function that has such a "half-prototype", specifying the return-type only:

1. Warn if you use such a function 
2. Do nothing; if the user wants to shoot oneself in the foot, let him do this. (gcc is such an example!)

The strategy no. 2 is good if you assume that the person who wrote this actually meant to have a void parameter list (for example, because he is used to C++, where this is handled differently).

Why is this difference important here?

The calling conventions at https://cc65.github.io/doc/cc65-intern.html tell us that the called function has to clean up the stack from the parameters. If there is a function (void), then there is nothing to do.

But if it is a variadic function (and a function without prototype must be treated as variadic!), then: 

> [...] the Y register contains the number of bytes pushed to the stack on entry, which may be added to sp to restore its original state.

Thus, this mismatch will clobber the stack and result in a failing program!

You might be used by other compilers that this is not a problem; however, this is an implementation detail of the compilers, because they use a calling convention that allows for it. For example, an x86/x64, each function has its stack frame with EBP. That is, if a called function behaves erroneously (like here), the calling function will not break, because it can still find everything with its EBP. In fact, when it returns itself, it will also clean up the mess from the called function, because it restores the ESP to what it was when it was called. It is kind of "self-healing".

The calling convention of cc65 does not allow this, though.

So, IMHO, a warning or error must be output here. The text might be misleading and might be improved, though. What about "function prototype mismatch" or something like this?

**Comment by polluks** [comment]
@spiro-trikaliotis Maybe we should include your statement.

**Comment by Marrin** [comment]
@neilbalch Sidenote: There are definitions of `SEI` and `CLI` in the `6502.h` header.  And with `NOP` you have to make sure to not compile this with optimizations because I would not expect them to survive the peep hole optimizer.

**Comment by neilbalch** [comment]
> 
> 
> @neilbalch Sidenote: There are definitions of `SEI` and `CLI` in the `6502.h` header. And with `NOP` you have to make sure to not compile this with optimizations because I would not expect them to survive the peep hole optimizer.

Oh, okay! Good to know, thanks

**Comment by mrdudz** [comment]
> > Excuse me, but the error message is still misleading. no matter what the implementation change was, a prototype that looks exactly like the function signature should work. It's also valid C no matter what.
> 
> Well, in C terms, void a(); is not a prototype.
> 
> Citing N1256 (the latest c99 draft), "6.2.1 scope of identifiers", No. 2: "A function prototype is a declaration of a function that declares the types of its parameters."
> 
> In contrast, the "void a();" does not define the type of the parameters. This it is not a prototype.
[snip]

Is it the same for other standards cc65 supports? (we dont actually support C99 :))

> So, IMHO, a warning or error must be output here. The text might be misleading and might be improved, though. What about "function prototype mismatch" or something like this?

Warning sounds reasonable, error - i dont know. While ugly, its a kindof common construct,

For curiosity, if this thing is technically not a prototype - what _is_ it? :)

**Comment by spiro-trikaliotis** [comment]
> Is it the same for other standards cc65 supports? (we dont actually support C99 :))

At least for pre-C99: yes. This might have changed later, I did not follow the later standards anymore.

>Warning sounds reasonable, error - i dont know. While ugly, its a kindof common construct,

Yes, it is a common construct, and many people will believe that this is valid (and ignore the warning). However, it is very likely that their program will not work as expected.

From this point of view, I am more for an error, because it is a severe problem in the cc65 environment.

> For curiosity, if this thing is technically not a prototype - what is it? :)

It's a forward declaration of a function name. You tell the compiler not to complain about a function "a()" not being known when it is used, because it will be declared afterwards, or externally.

From my point of view, this behaviour (and allowing this behaviour) is a consequence of ANSI C still supporting K&R C, which did not have any prototypes. Without K&R C, I think the C standard would have chosen the C++ approach, in which an empty paranthesis is the same as "(void)".

**Comment by neilbalch** [comment]
This is a very interesting discussion, so thanks for taking the time to explain things here. I'm a relative C noob _(more familiar with the C++ way of doing things)_ and I really appreciate it!
